plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Cse.hs

```diff
@@ -0,0 +1,398 @@
+
+-------------------------------------------------------------------------------
+2. How does it work?
+-------------------------------------------------------------------------------
+
+We use the following example to explain how the implementation works:
+
+\x y -> (1+(2+x))
+        +
+        (case y [ (1+(2+x)) + (3+x)
+                , (2+x) + (3+x)
+                , 4+x
+                ]
+        )
+
+The implementation makes several passes on the given term. The first pass collects builtin
+arity information as described above.
+
+In the second pass, we assign a unique ID to each `LamAbs`, `Delay`, and each `Case` branch.
+Then, we annotate each subterm with a path, consisting of IDs encountered from the root
+to that subterm (not including itself).
+
+In the above example, the ID of `\x` is 0, the ID of `\y` is 1, and the IDs of the
+three case branches are 2, 3, 4 (the actual numbers don't matter, as long as they are unique).
+The path for the first `1+(2+x)` and the first `2+x` is "0.1"; the path for the second
+`1+(2+x)` and the second `2+x` is "0.1.2"; the path for `4+x` is "0.1.4".
+
+In the third pass, we calculate a count for each `(term, path)` pair, where `term` is a
+non-workfree term, and `path` is its path. If the same term has two paths, and one is an
+ancestor (i.e., prefix) of the other, we increment the count for the ancestor path in both
+instances.
+
+In the above example, there are three occurrences of `2+x`, whose paths are "0.1", "0.1.2"
+and "0.1.3", respectively. The first path is an ancestor of the latter two. Therefore,
+the count for `(2+x, "0.1")` is 3, while the count for `(2+x, "0.1.2")` and `(2+x, "0.1.3")`
+is 0. The following all have a count of 1: `(3+x, "0.1.2")`, `(3+x, "0.1.3")` and
+`(4+x, "0.1.4")`.
+
+Now, each `(term, path)` pair whose count is greater than 1 is a CSE candidate.
+In the above example, the CSE candidates are `(2+x, "0.1")` and `(1+(2+x), "0.1")`.
+Note that `3+x` is not a CSE candidate, because it has two paths, and neither has a count
+greater than 1. `2+` is also not a CSE candidate, because it is workfree.
+
+The CSE candidates are then processed in descending order of their `termSize`s. For each CSE
+candidate, we generate a fresh variable, create a LamAbs for it under its path, and substitute
+it for all occurrences in the original term whose paths are descendents (or self) of
+the candidate's path. The order is because a bigger expression may contain a small subexpression.
+
+In the above example, we first process CSE candidate `(1+(2+x), "0.1")`. We create a fresh
+variable `cse1` for it, perform substitution, and create a `LamAbs` under path "0.1" (i.e., around
+the body of `y`). After processing this CSE candidate, the original term becomes
+
+\x y -> (\cse1 -> cse1
+                  +
+                  (case y [ cse1 + (3+x)
+                          , (2+x) + (3+x)
+                          , 4+x
+                          ]
+        ) (1+(2+x))
```

Comment:
Yeah, this seems a little tricky. What if the new CSE term now includes `cse1`? The term below hoists `cse2` outside the scope of `cse1`! At least it should probably be inside. I notice the the examples are inconsistent: below there is an example where we do put `cse2` inside, which seems more correct.

I guess that if we do that it's safe, but the argument is non-trivial:
- Suppose we have two CSE candidates `cse1` and `cse2`, where `cse1` occurs within `cse2`.
- Then if we decide to hoist `cse1`, we must do so to the level indicated by the path to `cse2`, since `cse1` is a subexpression of `cse2`
- We thus in theory need to extend the path for `cse2` by one step for the new binding, but since it occurs right under the previous last step, we can just use the old path so long as we put the binding for `cse2` under the other CSE bindings.

I'm not even sure if that is correct. I hope we have some tests for this and we need a careful argument!

Reply:
> The term below hoists cse2 outside the scope of cse1! At least it should probably be inside. 

Seems correct to me? Not sure which term you're referring too. I've been ignoring indices in the names, only looking at semantics.

Reply:
> But then shouldn't we handle such lambdas the same way regardless of whether they were originally in the program or were added by CSE?

Good point. What I'm not sure is: whether CSE should handle this, or should we implement another optimization, that floats terms inside an immediately-applied lambda that don't depend on the lambda-bound variable, outside of the lambda (i.e., similar to full-laziness).

> Suppose we have two CSE candidates cse1 and cse2, where cse1 occurs within cse2.

That's why the CSE candidates are processed in descending order of size. If we process cse2 first, then there's no problem.

Reply:
> What I'm not sure is: whether CSE should handle this, or should we implement another optimization, that floats terms inside an immediately-applied lambda that don't depend on the lambda-bound variable, outside of the lambda (i.e., similar to full-laziness).

I don't believe it's going to fly. Plutus Core programs tend to have a long prefix of immediately applied lambdas, are you going to extract half a program outside those by picking every single subexpression that doesn't happen to reference some of the variables? Plus it'll create bindings for things that we don't really want to create bindings for, like what's the point of turning `(\x -> 1 + 2 + x) e` into `(\y -> (\x -> y + x) e) (1 + 2)` -- no work is saved and some evaluation cycles are wasted on creating a variable. And each such transformation only creates more immediately applied lambdas.

I _suspect_ that full-laziness-like transformations only help when you lift a subexpression past a non-applied lambda. CSE is different in that regard: it helps whenever a subexpression occurs more than one time, regardless of whether you're inside an applied lambda or not.

Reply:
> That's why the CSE candidates are processed in descending order of size. If we process cse2 first, then there's no problem.

That's a bit subtle and needs to be written down. I had to think about it for a while. Here's my argument:

By processing CSE candidates in size order, we also implicitly process them in dependency order: if we have two CSE candidates `cse1` and `cse2`, then we cannot end up in a situation where `cse2` would appear in `cse1`, because then the original `cse1` would have had to have been bigger than `cse2`.

I'm actually not sure this is a sound argument, because doing some rounds of CSE can change the size ordering. Can we end up with a situation like this?
1. `cse1` < `cse2` < `cse3`
2. Pull out `cse1`
3. `cse3` < `cse2`, and `cse3` appears a subexpression of `cse2`
4. Pull out `cse2`
5. Pull out `cse3`, break the RHS of `cse2`

Looking at that, in fact I think what we need to be safe is just:
1. Bindings for subsequent CSE candidates go inside the previous ones
2. We only substitute for a CSE candidate inside the scope of the binding
Which is much simpler and probably holds.

That still implies that this example is wrong and `cse2` should be bound inside the scope of `cse1`.

Reply:
@michaelpj I don't get your point at all. What do you think is the correct output for this example?

`cse2` is smaller than `cse1`. Shouldn't the smaller one be in the outer scope? I.e., you'd want

```
let cse2 = 2 + x in
let cse1 = 1 + cse2 in body
```

If you flip it you'd get

```
let cse1 = 1 + (2 + x) in
let cse2 = 2 + x in body
```

> Can we end up with a situation like this?

In your example you are processing CSE candidates in asencding order of size, while I'm proposing descending order.

Reply:
Okay, I just had it backwards. You're doing them in _descending_ order of size :facepalm: 

Then yes, you need to put each subsequent one outside the other ones. And putting the new ones outside should always be safe.